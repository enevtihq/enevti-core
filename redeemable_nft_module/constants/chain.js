const CHAIN = {
  NAME: "enevti",
};

module.exports = CHAIN;
