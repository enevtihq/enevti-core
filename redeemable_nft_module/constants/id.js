const REDEEMABLE_NFT_MODULE_ID = 1000;
const REDEEMABLE_NFT_CREATE_ONEKIND_NFT = 1;

module.exports = {
  REDEEMABLE_NFT_MODULE_ID,
  REDEEMABLE_NFT_CREATE_ONEKIND_NFT,
};
