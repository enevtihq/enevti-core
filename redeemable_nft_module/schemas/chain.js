const CHAIN_STATE_ALL_NFT = "redeemable_nft_chain:all_nft";
const CHAIN_STATE_NFT = "redeemable_nft_chain";
const CHAIN_STATE_ALL_CONTAINER = "redeemable_nft_container:all_container";
const CHAIN_STATE_CONTAINER = "redeemable_nft_container";
const CHAIN_STATE_NFTPACK = "redeemable_nft_pack";
const CHAIN_STATE_REDEEM_STATUS_MONITOR = "redeemable_nft_monitor";

const redeemableNFTSchema = {
  $id: "enevti/redeemable_nft/nft",
  type: "object",
  required: ["id", "serial", "name", "description", "createdOn", "data", "ownerAddress", "originAddress", "originChain", "NFTType", "utility", "redeem", "rarity", "journey", "value", "onSale", "royalty"],
  properties: {
    id: {
      dataType: "bytes",
      fieldNumber: 1,
    },
    serial: {
      dataType: "string",
      fieldNumber: 2,
    },
    name: {
      dataType: "string",
      fieldNumber: 3,
    },
    description: {
      dataType: "string",
      fieldNumber: 4,
    },
    createdOn: {
      dataType: "uint32",
      fieldNumber: 5,
    },
    data: {
      dataType: "string",
      fieldNumber: 6,
    },
    ownerAddress: {
      dataType: "bytes",
      fieldNumber: 7,
    },
    originAddress: {
      dataType: "bytes",
      fieldNumber: 8,
    },
    originChain: {
      dataType: "string",
      fieldNumber: 9,
    },
    NFTType: {
      dataType: "string",
      fieldNumber: 10,
    },
    utility: {
      dataType: "string",
      fieldNumber: 11,
    },
    redeem: {
      type: "object",
      required: ["parts", "partition", "count", "message", "status", "recurring", "time", "from", "until", "limit", "touched"],
      fieldNumber: 12,
      properties: {
        parts: {
          type: "array",
          fieldNumber: 1,
          items: {
            dataType: "bytes",
          },
        },
        partition: {
          dataType: "uint32",
          fieldNumber: 2,
        },
        count: {
          dataType: "uint32",
          fieldNumber: 3,
        },
        message: {
          type: "object",
          required: ["cipher", "nonce"],
          fieldNumber: 4,
          properties: {
            cipher: {
              dataType: "string",
              fieldNumber: 1,
            },
            nonce: {
              dataType: "string",
              fieldNumber: 2,
            },
          },
        },
        status: {
          dataType: "string",
          fieldNumber: 5,
        },
        recurring: {
          dataType: "string",
          fieldNumber: 6,
        },
        time: {
          type: "object",
          fieldNumber: 7,
          properties: {
            day: {
              dataType: "uint32",
              fieldNumber: 1,
            },
            date: {
              dataType: "uint32",
              fieldNumber: 2,
            },
            month: {
              dataType: "uint32",
              fieldNumber: 3,
            },
            year: {
              dataType: "uint32",
              fieldNumber: 4,
            },
          },
        },
        from: {
          type: "object",
          required: ["hour", "minute"],
          fieldNumber: 8,
          properties: {
            hour: {
              dataType: "uint32",
              fieldNumber: 1,
            },
            minute: {
              dataType: "uint32",
              fieldNumber: 2,
            },
          },
        },
        until: {
          dataType: "uint32",
          fieldNumber: 9,
        },
        limit: {
          dataType: "uint32",
          fieldNumber: 10,
        },
        touched: {
          dataType: "uint32",
          fieldNumber: 11,
        },
      },
    },
    rarity: {
      dataType: "uint32",
      fieldNumber: 13,
    },
    journey: {
      type: "array",
      fieldNumber: 14,
      items: {
        activity: {
          dataType: "string",
          fieldNumber: 1,
        },
        date: {
          dataType: "uint32",
          fieldNumber: 2,
        },
        by: {
          dataType: "bytes",
          fieldNumber: 3,
        },
        from: {
          dataType: "bytes",
          fieldNumber: 4,
        },
        message: {
          dataType: "string",
          fieldNumber: 5,
        },
        amount: {
          dataType: "uint64",
          fieldNumber: 6,
        },
      },
    },
    price: {
      type: "object",
      required: ["amount", "currency"],
      fieldNumber: 15,
      properties: {
        amount: {
          dataType: "uint64",
          fieldNumber: 1,
        },
        currency: {
          dataType: "string",
          fieldNumber: 2,
        },
      },
    },
    onSale: {
      dataType: "boolean",
      fieldNumber: 16,
    },
    royalty: {
      type: "object",
      properties: {
        origin: {
          dataType: "uint32",
          fieldNumber: 1,
        },
        staker: {
          dataType: "uin32",
          fieldNumber: 2,
        },
      },
    },
  },
};

const NFTContainerSchema = {
  $id: "enevti/redeemable_nft/nft_container",
  type: "object",
  required: ["id", "containerType", "allNFT", "NFTSold", "availableItem", "mintingExpire", "originAddress", "price", "name", "description", "packSize", "createdOn"],
  properties: {
    id: {
      dataType: "bytes",
      fieldNumber: 1,
    },
    containerType: {
      dataType: "string",
      fieldNumber: 2,
    },
    allNFT: {
      type: "array",
      fieldNumber: 3,
      items: {
        dataType: "bytes",
      },
    },
    NFTSold: {
      type: "array",
      fieldNumber: 4,
      items: {
        dataType: "bytes",
      },
    },
    availableItem: {
      type: "array",
      fieldNumber: 5,
      items: {
        dataType: "bytes",
      },
    },
    mintingExpire: {
      dataType: "uint32",
      fieldNumber: 6,
    },
    originAddress: {
      dataType: "bytes",
      fieldNumber: 7,
    },
    price: {
      type: "object",
      fieldNumber: 8,
      properties: {
        amount: {
          dataType: "uint64",
          fieldNumber: 1,
        },
        currency: {
          dataType: "string",
          fieldNumber: 2,
        },
      },
    },
    name: {
      dataType: "string",
      fieldNumber: 9,
    },
    description: {
      dataType: "string",
      fieldNumber: 10,
    },
    packSize: {
      dataType: "uint32",
      fieldNumber: 11,
    },
    createdOn: {
      dataType: "uint32",
      fieldNumber: 12,
    },
  },
};

const NFTPackSchema = {
  $id: "enevti/redeemable_nft/nft_pack",
  type: "object",
  required: ["id", "NFTItem", "parentContainer", "boughtBy"],
  properties: {
    id: {
      dataType: "bytes",
      fieldNumber: 1,
    },
    NFTItem: {
      type: "array",
      fieldNumber: 2,
      items: {
        dataType: "bytes",
      },
    },
    parentContainer: {
      dataType: "bytes",
      fieldNumber: 3,
    },
    boughtBy: {
      dataType: "bytes",
      fieldNumber: 4,
    },
  },
};

const allNFTSchema = {
  $id: "enevti/redeemable_nft/all_nft",
  type: "object",
  required: ["allNFT"],
  properties: {
    allNFT: {
      type: "array",
      fieldNumber: 1,
      items: {
        dataType: "bytes",
      },
    },
  },
};

const allNFTContainerSchema = {
  $id: "enevti/redeemable_nft/all_nft_container",
  type: "object",
  required: ["allNFTContainer"],
  properties: {
    allNFTContainer: {
      type: "array",
      fieldNumber: 1,
      items: {
        dataType: "bytes",
      },
    },
  },
};

const redeemMonitorSchema = {
  $id: "enevti/redeemable_nft/redeem_monitor",
  type: "object",
  required: ["checkpoint", "all"],
  properties: {
    checkpoint: {
      type: "object",
      required: ["time", "nft"],
      fieldNumber: 1,
      properties: {
        time: {
          dataType: "uint32",
          fieldNumber: 1,
        },
        nft: {
          type: "array",
          fieldNumber: 2,
          items: {
            dataType: "bytes",
          },
        },
      },
    },
    all: {
      type: "array",
      fieldNumber: 2,
      items: redeemableNFTSchema,
    },
  },
};

module.exports = {
  redeemableNFTSchema,
  NFTContainerSchema,
  NFTPackSchema,
  allNFTSchema,
  allNFTContainerSchema,
  redeemMonitorSchema,
  CHAIN_STATE_ALL_NFT,
  CHAIN_STATE_NFT,
  CHAIN_STATE_ALL_CONTAINER,
  CHAIN_STATE_CONTAINER,
  CHAIN_STATE_NFTPACK,
  CHAIN_STATE_REDEEM_STATUS_MONITOR,
};
