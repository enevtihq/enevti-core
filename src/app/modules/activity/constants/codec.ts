export const ACTIVITY_PREFIX = 'activity';
export const GENESIS_ACTIVITY_PREFIX = 'genesis';
