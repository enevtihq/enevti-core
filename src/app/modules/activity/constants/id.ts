export const ACTIVITY_MODULE_ID = 1005;
