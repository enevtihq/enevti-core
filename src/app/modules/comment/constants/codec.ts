export const COMMENT_PREFIX = 'comment';
export const ADD_COMMENT_ASSET_NAME = 'addComment';
