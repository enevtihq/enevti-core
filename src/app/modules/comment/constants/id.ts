export const COMMENT_MODULE_ID = 1008;
export const ADD_COMMENT_ASSET_ID = 0;
