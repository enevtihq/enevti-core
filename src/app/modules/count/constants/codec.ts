export const COUNT_PREFIX = 'count';
export const COUNT_ITEM_PREFIX = 'item';
