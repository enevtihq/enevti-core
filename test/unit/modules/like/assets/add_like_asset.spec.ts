import { AddLikeProps } from 'enevti-types/asset/like/add_like_asset';
import { LikeChain, LikedChain } from 'enevti-types/chain/like';
import { codec, ReducerHandler, StateStore, testing } from 'lisk-sdk';
import { COUNT_PREFIX } from '../../../../../src/app/modules/count/constants/codec';
import { AddLikeAsset } from '../../../../../src/app/modules/like/assets/add_like_asset';
import {
  ADD_LIKE_ASSET_NAME,
  LIKED_PREFIX,
  LIKE_MODULE_PREFIX,
  LIKE_PREFIX,
} from '../../../../../src/app/modules/like/constants/codec';
import { ADD_LIKE_ASSET_ID } from '../../../../../src/app/modules/like/constants/id';
import {
  IDENTIFIER_MAX_LENGTH,
  ID_MAX_LENGTH,
} from '../../../../../src/app/modules/like/constants/limit';
import { addLikeSchema } from '../../../../../src/app/modules/like/schema/assets/add_like_assets';
import { likeSchema } from '../../../../../src/app/modules/like/schema/like';
import { likedSchema } from '../../../../../src/app/modules/like/schema/liked';
import { getLike } from '../../../../../src/app/modules/like/utils/like';
import { getLiked } from '../../../../../src/app/modules/like/utils/liked';

describe('AddLikeAsset', () => {
  let transactionAsset: AddLikeAsset;

  beforeEach(() => {
    transactionAsset = new AddLikeAsset();
  });

  describe('constructor', () => {
    it('should have valid id', () => {
      expect(transactionAsset.id).toEqual(ADD_LIKE_ASSET_ID);
    });

    it('should have valid name', () => {
      expect(transactionAsset.name).toEqual(ADD_LIKE_ASSET_NAME);
    });

    it('should have valid schema', () => {
      expect(transactionAsset.schema).toEqual(addLikeSchema);
    });
  });

  describe('validate', () => {
    describe('schema validation', () => {
      it('should throw error if identifier length exceeding limit', () => {
        const context = testing.createValidateAssetContext<AddLikeProps>({
          asset: { identifier: 'a'.repeat(IDENTIFIER_MAX_LENGTH + 1), id: 'id' },
          transaction: { senderAddress: Buffer.alloc(0) } as any,
        });
        expect(() => transactionAsset.validate(context)).toThrow();
      });

      it('should throw error if id length exceeding limit', () => {
        const context = testing.createValidateAssetContext<AddLikeProps>({
          asset: { identifier: 'identifier', id: 'a'.repeat(ID_MAX_LENGTH + 1) },
          transaction: { senderAddress: Buffer.alloc(0) } as any,
        });
        expect(() => transactionAsset.validate(context)).toThrow();
      });

      it('should be ok for valid schema', () => {
        const context = testing.createValidateAssetContext<AddLikeProps>({
          asset: { id: 'id', identifier: 'identifier' },
          transaction: { senderAddress: Buffer.alloc(0) } as any,
        });
        expect(() => transactionAsset.validate(context)).not.toThrow();
      });
    });
  });

  describe('apply', () => {
    let stateStore: StateStore;
    let context;

    const reducerHandler: ReducerHandler = testing.mocks.reducerHandlerMock;
    const identifier = 'identifier';
    const target = 'target';
    const { address } = testing.fixtures.defaultFaucetAccount;
    const likeChain: LikeChain = {
      address: [address],
    };
    const likedChain: LikedChain = {
      status: 1,
    };
    const likeChainValue = codec.encode(likeSchema, likeChain);
    const likedChainValue = codec.encode(likedSchema, likedChain);

    beforeEach(() => {
      const chain = {
        [`${LIKE_MODULE_PREFIX}:${identifier}:${target}:${LIKE_PREFIX}`]: likeChainValue,
        [`${LIKE_MODULE_PREFIX}:${LIKED_PREFIX}:${target}:${address.toString(
          'hex',
        )}`]: likedChainValue,
      };

      stateStore = new testing.mocks.StateStoreMock({
        chain,
      });

      context = testing.createApplyAssetContext<AddLikeProps>({
        stateStore,
        reducerHandler,
        asset: { identifier, id: target },
        transaction: { senderAddress: address, nonce: BigInt(1) } as any,
      });

      jest.spyOn(reducerHandler, 'invoke');
      jest.spyOn(stateStore.chain, 'get');
      jest.spyOn(stateStore.chain, 'set');
    });

    it('should add like to blockchain state', async () => {
      context = testing.createApplyAssetContext<AddLikeProps>({
        stateStore,
        reducerHandler,
        asset: { identifier: 'newIdentifier', id: 'newTarget' },
        transaction: { senderAddress: address, nonce: BigInt(1) } as any,
      });

      await transactionAsset.apply(context);

      const like = await getLike(stateStore, context.asset.identifier, context.asset.id);
      expect(like).toEqual(likeChain);
    });

    it('should throw an error if sender already give a like', async () => {
      const func = async () => {
        try {
          await transactionAsset.apply(context);
          return true;
        } catch {
          return false;
        }
      };
      expect(await func()).toBe(false);
    });

    it('should set liked to 1 if sender has not been give a like', async () => {
      context = testing.createApplyAssetContext<AddLikeProps>({
        stateStore,
        reducerHandler,
        asset: { identifier: 'newIdentifier', id: 'newTarget' },
        transaction: { senderAddress: address, nonce: BigInt(1) } as any,
      });

      await transactionAsset.apply(context);

      const liked = await getLiked(stateStore, context.asset.id, context.transaction.senderAddress);
      expect(liked).toEqual({ status: 1 });
    });

    it('should unshift new like target data to blockchain', async () => {
      const senderAddress = Buffer.from('newAddress', 'utf-8');
      context = testing.createApplyAssetContext<AddLikeProps>({
        stateStore,
        reducerHandler,
        asset: { identifier, id: target },
        transaction: { senderAddress, nonce: BigInt(1) } as any,
      });

      await transactionAsset.apply(context);

      const like = await getLike(stateStore, context.asset.identifier, context.asset.id);
      expect(like).toEqual({ address: [senderAddress, ...likeChain.address] });
    });

    it('should only contain new target data for the first time', async () => {
      const senderAddress = Buffer.from('newAddress', 'utf-8');
      context = testing.createApplyAssetContext<AddLikeProps>({
        stateStore,
        reducerHandler,
        asset: { identifier: 'newIdentifier', id: 'newTarget' },
        transaction: { senderAddress, nonce: BigInt(1) } as any,
      });

      await transactionAsset.apply(context);

      const like = await getLike(stateStore, context.asset.identifier, context.asset.id);

      expect(like).toEqual({ address: [senderAddress] });
    });

    it('should invoke count module addCount reducers for succesful add like operation', async () => {
      context = testing.createApplyAssetContext<AddLikeProps>({
        stateStore,
        reducerHandler,
        asset: { identifier: 'newIdentifier', id: 'newTarget' },
        transaction: { senderAddress: address, nonce: BigInt(1) } as any,
      });

      await transactionAsset.apply(context);

      expect(reducerHandler.invoke).toHaveBeenCalledWith(`${COUNT_PREFIX}:addCount`, {
        module: LIKE_MODULE_PREFIX,
        key: 'newIdentifier',
        address,
        item: Buffer.from('newTarget', 'hex'),
      });
    });
  });
});
