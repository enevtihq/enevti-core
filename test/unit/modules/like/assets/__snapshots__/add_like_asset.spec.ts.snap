// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddLikeAsset constructor should have valid schema 1`] = `
Object {
  "$id": "enevti/like/addLike",
  "properties": Object {
    "id": Object {
      "dataType": "string",
      "fieldNumber": 2,
    },
    "identifier": Object {
      "dataType": "string",
      "fieldNumber": 1,
    },
  },
  "required": Array [
    "identifier",
    "id",
  ],
  "title": "addLikeAsset transaction asset for like module",
  "type": "object",
}
`;
